lib/ymodel/lib/ymodel/loadable.rb

```diff
@@ -21,11 +21,40 @@ def schema
       YModel::Schema.new({})
     end
 
+    def load_records!
+      all = records.map { |record| new(record) }
+      unless all.map(&:index) == all.map(&:index).uniq
+        raise YModel::DuplicateIndexError, 'Some records share the same index.'
+      end
+
+      all
+    end
+
+    def index_on(key)
+      @index = key
+    end
+
+    def index
+      @index || :id
+    end
+
     protected
 
+    # I dislike the format of this method a lot. Rubocop made me do it..
     def records
-      @records ||= YAML.load_file(source)
-        .map(&:symbolize_keys)
+      @records ||=
+        if compiled?
+          Dir.foreach(source).reduce([]) do |memo, filename|
+            if filename.match?(/\A.*(.yaml|.yml)\z/)
+              memo + YAML.load_file(File.join(source, filename))
+            else
+              memo
+            end
+          end
```

Comment:
Personally, I don't think we should allow both ".yaml" and ".yml" extensions; it's reasonable to expect files to have a consistent naming convention. Elsewhere in ETModel (and other projects) it's always ".yml". The `_source` method for non-compiled YModel files still expects only ".yml".

```suggestion
          Dir.glob(File.join(source, '*.yml'))
            .flat_map { |name| YAML.load_file(name) }
```